cmd/kubeadm/app/phases/controlplane/components.go

```diff
@@ -0,0 +1,174 @@
+/*
+Copyright 2023 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package controlplane
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"io"
+	"os"
+	"path/filepath"
+	"time"
+
+	"github.com/pkg/errors"
+	"gopkg.in/yaml.v2"
+
+	v1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/util/wait"
+	"k8s.io/client-go/rest"
+	"k8s.io/klog/v2"
+	kubeletconfig "k8s.io/kubelet/config/v1beta1"
+
+	kubeadmconstants "k8s.io/kubernetes/cmd/kubeadm/app/constants"
+	"k8s.io/kubernetes/cmd/kubeadm/app/util/staticpod"
+)
+
+var ControlPlaneComponents = []string{
+	kubeadmconstants.KubeAPIServer,
+	kubeadmconstants.KubeControllerManager,
+	kubeadmconstants.KubeScheduler,
+}
+
+type component struct {
+	name    string
+	labels  map[string]string
+	touched bool
+}
+
+// WaitForControlPlaneComponent wait for control plane component to be ready by check pod status returned by kubelet
+func WaitForControlPlaneComponents(componentNames []string, timeout time.Duration, manifestDir, kubeletDir, certificatesDir string) error {
+	certFile := filepath.Join(certificatesDir, kubeadmconstants.APIServerKubeletClientCertName)
+	keyFile := filepath.Join(certificatesDir, kubeadmconstants.APIServerKubeletClientKeyName)
```

Comment:
can we use the cert and key for the kubelet client stored in /etc/kubernetes/kubelet.conf (they link to files under /var/lib/kubelet...)?

Reply:
Do you mean file `/var/lib/kubelet/pki/kubelet-client-current.pem`  ?
When I try to run cmd: 
``` shell
curl -k --cert /var/lib/kubelet/pki/kubelet-client-current.pem --key /var/lib/kubelet/pki/kubelet-client-current.pem https://127.0.0.1:10250/pods
```
got the following result:
`Forbidden (user=system:node:localhost.localdomain, verb=get, resource=nodes, subresource=proxy)`

By the way, why cert and key of apiserver-kubelet-client is not good to fetch pods info from kubelet?

Reply:
i was wondering if we can just use the less privileged kubelet cert. apparently not. let's continue with the apiserver cert for now.

thanks